keyratelimit.go

```diff
@@ -43,68 +42,72 @@ func (m *MultiLimiter) Add(opts *Options) error {
 	if err := opts.Validate(); err != nil {
 		return err
 	}
-	_, ok := m.limiters[opts.Key]
-	if ok {
-		return fmt.Errorf("key already exists")
-	}
 	var rlimiter *Limiter
 	if opts.IsUnlimited {
 		rlimiter = NewUnlimited(m.ctx)
 	} else {
 		rlimiter = New(m.ctx, opts.MaxCount, opts.Duration)
 	}
-	m.limiters[opts.Key] = rlimiter
+	// ok if true if key already exists
+	_, ok := m.limiters.LoadOrStore(opts.Key, rlimiter)
+	if ok {
+		return fmt.Errorf("key already exists")
+	}
 	return nil
 }
 
 // GetLimit returns current ratelimit of given key
 func (m *MultiLimiter) GetLimit(key string) (uint, error) {
-	limiter, ok := m.limiters[key]
-	if !ok || limiter == nil {
-		return 0, fmt.Errorf("key doesnot exist")
+	limiter, err := m.get(key)
+	if err != nil {
+		return 0, err
 	}
 	return limiter.GetLimit(), nil
 }
 
 // Take one token from bucket returns error if key not present
 func (m *MultiLimiter) Take(key string) error {
-	limiter, ok := m.limiters[key]
-	if !ok || limiter == nil {
-		return fmt.Errorf("key doesnot exist")
+	limiter, err := m.get(key)
+	if err != nil {
+		return err
 	}
 	limiter.Take()
 	return nil
 }
 
-// Stop internal limiters with defined keys or all if no key is provided
-func (m *MultiLimiter) Stop(keys ...string) {
-	if len(keys) > 0 {
-		m.stopWithKeys(keys...)
-	} else {
-		m.stopWithKeys(maps.Keys(m.limiters)...)
+// AddAndTake adds key if not present and then takes token from bucket
+func (m *MultiLimiter) AddAndTake(opts *Options) {
+	if limiter, err := m.get(opts.Key); err == nil {
+		limiter.Take()
+		return
 	}
+	_ = m.Add(opts)
+	_ = m.Take(opts.Key)
 }
 
-// stopWithKeys stops the internal limiters matching keys
-func (m *MultiLimiter) stopWithKeys(keys ...string) {
-	for _, key := range keys {
-		if limiter, ok := m.limiters[key]; ok {
+// Stop internal limiters with defined keys or all if no key is provided
+func (m *MultiLimiter) Stop(keys ...string) {
+	if len(keys) == 0 {
+		m.limiters.Range(func(key, value any) bool {
+			value.(*Limiter).Stop()
+			return true
+		})
+		return
+	}
+	for _, v := range keys {
+		if limiter, err := m.get(v); err == nil {
 			limiter.Stop()
 		}
 	}
 }
 
-// SleepandReset stops timer removes all tokens and resets with new limit (used for Adaptive Ratelimiting)
-func (m *MultiLimiter) SleepandReset(SleepTime time.Duration, opts *Options) error {
-	if err := opts.Validate(); err != nil {
-		return err
+// get returns *Limiter instance
+func (m *MultiLimiter) get(key string) (*Limiter, error) {
+	val, _ := m.limiters.Load(key)
+	if val == nil {
+		return nil, fmt.Errorf("multilimiter: key does not exist")
```

Comment:
Unless we need an error with format, we can use a generic defined one as suggested before